scheduler/pkg/server/pipeline_status_test.go

```diff
@@ -210,3 +211,276 @@ func TestPipelineStatusEvents(t *testing.T) {
+			s.pipelineGwRebalance()
+			time.Sleep(500 * time.Millisecond)
+
+			var msr *pb.PipelineStatusResponse
+			select {
+			case next := <-stream.msgs:
+				msr = next
+			default:
+				t.Fail()
+			}
+
+			g.Expect(msr).ToNot(BeNil())
+			g.Expect(msr.PipelineName).To(Equal(test.loadReq.Name))
+			g.Expect(msr.Versions).To(HaveLen(1))
+		})
+	}
+}
+
+func TestPipelineGwRebalance(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	type test struct {
+		name      string
+		pipelines []*pb.Pipeline
+		replicas  int // number of pipelinegw instances
+	}
+
+	tests := []test{
+		{
+			name: "rebalance multiple models across N replicas",
+			pipelines: []*pb.Pipeline{
+				{
+					Name:    "pipeline1",
+					Version: 1,
+					Uid:     "uid1",
+					Steps: []*pb.PipelineStep{
+						{
+							Name: "step1",
+						},
+						{
+							Name:   "step2",
+							Inputs: []string{"step1.outputs"},
+						},
+					},
+				},
+				{
+					Name:    "pipeline2",
+					Version: 1,
+					Uid:     "uid2",
+					Steps: []*pb.PipelineStep{
+						{
+							Name: "step1",
+						},
+						{
+							Name:   "step2",
+							Inputs: []string{"step1.outputs"},
+						},
+					},
+				},
+				{
+					Name:    "pipeline3",
+					Version: 1,
+					Uid:     "uid3",
+					Steps: []*pb.PipelineStep{
+						{
+							Name: "step1",
+						},
+						{
+							Name:   "step2",
+							Inputs: []string{"step1.outputs"},
+						},
+					},
+				},
+			},
+			replicas: 4, // test with 4 modelgw replicas
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			s, _ := createTestScheduler()
+
+			var streams []*stubPipelineStatusServer
+			for i := 0; i < test.replicas; i++ {
+				// create pipelinegw stream
+				stream := newStubPipelineStatusServer(20, 5*time.Millisecond)
+				subscription := &PipelineSubscription{
+					name:              fmt.Sprintf("dummy%d", i),
+					ip:                fmt.Sprintf("127.0.0.%d", i+1),
+					isPipelineGateway: true,
+					stream:            stream,
+					fin:               make(chan bool),
+				}
+				s.pipelineEventStream.streams[stream] = subscription
+				s.pipelineEventStream.namesToIps[subscription.name] = fmt.Sprintf("127.0.0.%d", i+1)
+				g.Expect(s.pipelineEventStream.streams[stream]).ToNot(BeNil())
+
+				// add stream to the load balancer
+				s.pipelineGWLoadBalancer.AddServer(subscription.name)
+				streams = append(streams, stream)
+			}
+			g.Expect(len(s.pipelineEventStream.streams)).To(Equal(test.replicas))
+
+			// Load all pipelines into the store
+			for _, req := range test.pipelines {
+				err := s.pipelineHandler.AddPipeline(req)
+				g.Expect(err).To(BeNil())
+			}
+
+			// Allow events to propagate
+			time.Sleep(500 * time.Millisecond)
+
+			// Read all messages from the streams before rebalance
+			for i, stream := range streams {
+			drainLoop:
+				for {
+					select {
+					case msg := <-stream.msgs:
+						log.Infof("Drained message from stream %d: %v", i, msg)
+					default:
```

Comment:
Similar to comment above regarding allowing for a longer event propagation time but without having all of it as a prepended sleep.